crypt_keystore.go

```diff
@@ -0,0 +1,166 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2019 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package secboot
+
+import (
+	"errors"
+
+	"golang.org/x/xerrors"
+)
+
+// ErrUnsupportedKeyHandle is returned from a KeyStore that doesn't handle the supplied KeyHandle type.
+var ErrUnsupportedKeyHandle = errors.New("the supplied key handle is not supported by the key store")
+
+// KeyRecoverErrorCode describes an error returned from KeyStore.Recover.
+type KeyRecoverErrorCode int
+
+const (
+	// KeyRecoverUnexpectedError indicates an error that was unexpected and doesn't have a more appropriate error code.
+	KeyRecoverUnexpectedError KeyRecoverErrorCode = iota + 1
+
+	// KeyRecoverKeyStoreTemporarilyUnavailableError indicates that the keystore backend is temporarily unable to recover
+	// a key. In the case of a TPM, this could be because the TPM's dictionary attack protection has been triggered.
+	KeyRecoverKeyStoreTemporarilyUnavailableError
+
+	// KeyRecoverKeyStoreResourcesUnavailableError indicates that a key store doesn't contain the resources required to
+	// recover a key. In the case of a TPM, this could be because the TPM is not correctly provisioned.
+	KeyRecoverKeyStoreResourcesUnavailableError
+
+	// KeyRecoverInvalidKeyDataError indicates that a key store couldn't recover a key because the data provided with the
+	// key handle was invalid.
+	KeyRecoverInvalidKeyDataError
+
+	// KeyRecoverPINFailError indicates that a key store couldn't recover a key because the supplied PIN was incorrect.
+	KeyRecoverPINFailError
+)
+
+// KeyRecoverError is returned from KeyStore implementations to indicate the reason that a key could not be recovered.
+type KeyRecoverError struct {
+	Code KeyRecoverErrorCode
+	err  error
+}
+
+// WrapKeyRecoverError wraps the supplied error with a KeyRecoverError with the supplied error code. The err argument can be nil
+// if there is no need to provide any additional context.
+func WrapKeyRecoverError(code KeyRecoverErrorCode, err error) error {
+	return &KeyRecoverError{Code: code, err: err}
+}
+
+func (e *KeyRecoverError) Error() string {
+	var prefix string
+	switch e.Code {
+	case KeyRecoverUnexpectedError:
+		prefix = "unexpected error"
+	case KeyRecoverKeyStoreTemporarilyUnavailableError:
+		prefix = "the keystore is temporarily unavailable for recovering keys"
```

Comment:
can these errors escape to the user? it's a bit unclear whether surfacing the concept of keystore externally in errors makes sense